AGENTS.md

```diff
@@ -516,6 +401,300 @@ export const MyResource = Resource(
+    }
+  }
+  return this.destroy();
+}
+```
+
+#### Internal API Types Convention
+
+Mark internal types used only for API serialization with JSDoc `@internal`:
+
+```ts
+/**
+ * Raw provider API response for resource creation
+ * @internal
+ */
+interface ResourceApiResponse {
+  id: string;
+  name: string;
+  created_at: number;
+  // API field names may differ from output
+}
+```
+
+#### Retry Patterns with Exponential Backoff
+
+Use exponential backoff for transient errors:
+
+```ts
+import { withExponentialBackoff } from "../util/retry.ts";
+
+const result = await withExponentialBackoff(
+  async () => {
+    return await extractProviderResult<ApiResponse>(
+      `create resource "${name}"`,
+      api.post("/resources", requestBody)
+    );
+  },
+  (error) => {
+    // Retry condition: specific transient errors
+    return error.code === 1002 || error instanceof TimeoutError;
+  },
+  30,    // maximum attempts
+  100,   // initial delay in ms
+);
+```
+
+### Type Definition Patterns
+
+Alchemy resources follow a specific type definition pattern that ensures type safety and consistency. The key principle is that **the output interface name MUST match the exported resource name** to create a pseudo-class construct:
+
+#### Flat Properties vs Nested Objects
+
+Prefer flat properties over nested configuration objects for better developer experience and type safety:
+
+```ts
+// ✅ PREFERRED: Flat properties
+export interface MyResourceProps {
+  name?: string;
+  region: string;
+  secret?: string | Secret;
+  timeout?: number;
+}
+
+// ❌ AVOID: Nested configuration objects
+export interface MyResourceProps {
+  name?: string;
+  config: {
+    region: string;
+    secret?: string | Secret;
+    timeout?: number;
+  };
+}
+```
+
+Flat properties provide:
+- Better IDE autocomplete and type checking
+- Cleaner resource creation syntax
+- Easier validation and error handling
+- More intuitive API design
+
+```ts
+// ✅ CORRECT: Interface name matches exported resource name
+export type MyResource = {
+  // ... properties
+}
+export const MyResource = Resource(/* ... */);
+
+// ❌ INCORRECT: Interface name doesn't match
+export interface MyResourceOutput extends Resource<"provider::MyResource"> {
+  // ... properties
+}
+export const MyResource = Resource(/* ... */);
+```
+
+#### Props Interface Definition
+
+Define Props interface for input parameters:
+
+```ts
+export interface MyResourceProps {
+  /**
+   * Name of the resource
+   * @default ${app}-${stage}-${id}
+   */
+  name?: string;
+
+  /**
+   * Property description
+   */
+  property: string;
+
+  /**
+   * Secret value for authentication
+   * Use alchemy.secret() to securely store this value
+   */
+  secret?: string | Secret;
+
+  /**
+   * Whether to adopt an existing resource
+   * @default false
+   */
+  adopt?: boolean;
+
+  // Internal properties for lifecycle management
+  /** @internal */
+  resourceId?: string;
```

Comment:
internal doesn't make much sense being here, if we expose it to the user it isn't internal